app.go

```diff
@@ -299,6 +308,108 @@ func NewApp(
 	return app, nil
 }
 
+// InitAppForTestnet executes the necessary state transitions on the
+// provided application in order to start an in-place testnet.
+func InitAppForTestnet(app *App, pubKey crypto.PubKey, rawConsensusAddress bytes.HexBytes, rawOperatorAddress string, upgradeToTrigger string) *App {
+	ctx := app.NewUncachedContext(true, cmtproto.Header{})
+
+	// ===== Cosmos SDK: Staking  =====
+
+	cmtPubKey := &ed25519.PubKey{Key: pubKey.Bytes()}
+	operatorAddress := sdk.MustBech32ifyAddressBytes("noblevaloper", sdk.MustAccAddressFromBech32(rawOperatorAddress))
+	validator, err := stakingtypes.NewValidator(
+		operatorAddress,
+		cmtPubKey,
+		stakingtypes.Description{Moniker: "Testnet Validator"},
+	)
+	if err != nil {
+		cmtos.Exit("failed to create new validator: " + err.Error())
+	}
+
+	validator.Status = stakingtypes.Bonded
+	validator.Tokens = math.NewInt(1000000)
+	validator.DelegatorShares = math.LegacyNewDec(1000000)
+
+	stakingStore := ctx.KVStore(app.GetKey(stakingtypes.ModuleName))
+
+	// Remove all validators from last validators store
+	iterator, err := app.StakingKeeper.LastValidatorsIterator(ctx)
+	if err != nil {
+		cmtos.Exit("failed to create last validators iterator: " + err.Error())
+	}
+	for ; iterator.Valid(); iterator.Next() {
+		stakingStore.Delete(iterator.Key())
+	}
+	iterator.Close()
+
+	// Remove all validators from power store
+	iterator, err = app.StakingKeeper.ValidatorsPowerStoreIterator(ctx)
+	if err != nil {
+		cmtos.Exit("failed to create validators power store iterator: " + err.Error())
+	}
+	for ; iterator.Valid(); iterator.Next() {
+		stakingStore.Delete(iterator.Key())
+	}
+	iterator.Close()
+
+	// Remove all validators from validators store
+	iterator = storetypes.KVStorePrefixIterator(stakingStore, stakingtypes.ValidatorsKey)
+	for ; iterator.Valid(); iterator.Next() {
+		stakingStore.Delete(iterator.Key())
+	}
+	iterator.Close()
+
+	// Add our validator to power and last validators store
+	err = app.StakingKeeper.SetValidator(ctx, validator)
+	if err != nil {
+		cmtos.Exit("failed to set validator: " + err.Error())
+	}
+	err = app.StakingKeeper.SetValidatorByConsAddr(ctx, validator)
+	if err != nil {
+		cmtos.Exit("failed to set validator by consensus address: " + err.Error())
+	}
+	err = app.StakingKeeper.SetValidatorByPowerIndex(ctx, validator)
+	if err != nil {
+		cmtos.Exit("failed to set validator by power index: " + err.Error())
+	}
+	valAddress, _ := sdk.ValAddressFromBech32(validator.GetOperator())
+	err = app.StakingKeeper.SetLastValidatorPower(ctx, valAddress, 0)
+	if err != nil {
+		cmtos.Exit("failed to set last validator power: " + err.Error())
+	}
+	if err := app.StakingKeeper.Hooks().AfterValidatorCreated(ctx, valAddress); err != nil {
+		panic(err)
```

Comment:
is there any reason we use `cmtos.Exit` calls in a bunch of places where we encounter an non-recoverable error but here we use a `panic` instead?

Reply:
No reason! Updated 😄